database/mongodb/mongodb.go

```diff
@@ -184,10 +220,96 @@ func (m *Mongo) Drop() error {
 	return m.db.Drop(context.TODO())
 }
 
-func (m *Mongo) Lock() error {
+// Note that this could possibly have a race condition
```

Comment:
Thanks for pointing this out!
It may not be an issue in MongoDB 4.2: https://docs.mongodb.com/manual/reference/method/db.collection.createIndex/#concurrency
Also, can this be avoided by using strict [read/write concerns](https://pkg.go.dev/go.mongodb.org/mongo-driver@v1.4.1/mongo/options#CollectionOptions)?

Reply:
So this comment is slightly misleading (my fault), it was based on our evaluation of Mongo's behavior with 3.6 prior to deciding to implementing locking in go-migrate. Based on what you posted, my read is actually this actually might be an issue on 4.2 and newer.

Imagine we have three processes executing this A, B and C. 

They all make calls to createIndex() which [the docs say is allowed](https://docs.mongodb.com/manual/reference/method/db.collection.createIndex/#recreating-an-existing-index).  What I've observed in Mongo 3.6 is that connection A will build the index and B and C will simply return right way as if successful. If this behaviour is the general case, then in Mongo 4.0 and earlier, connection A will build the index, and B, and C will simply return right away.  B and C will then attempt to insert into the collection, which will block because the index has an exclusive lock and we aren't using `background`. 

However Mongo 4.2 ignores the `background` option, and doesn't hold a lock. So I think with Mongo 4.2 and newer, you can simply have B and C start trying to insert the lock right away without the index existing. In theory if this happened B and C would insert duplicate rows, and then the createIndex call would fail for connection A, because the unique constraint failed.

I suspect it would be hard for us to reproduce this issue in a real use case (we can test this easily by simply manually creating this collection (without an index), inserting 10,000,000 rows in it, and then trying to create this index on multiple connections in parallel), and we _might_ never run into it because the create Index call on an empty collection should be almost instant. However it is not concurrent safe as is, and the question is whether we should in your opinion @dhui.

I think we could mitigate this by explicitly checking if the index exists on the collection, when I was testing this behaviour the second connection didn't see the index when it inspected the collection.

Reply:
My understanding is that [MongoDB will lock](https://docs.mongodb.com/manual/core/index-creation/#index-build-process) (either at the collection or database level, depending on the version) to create an index. Assuming `getCollection()` calls `createCollection()`, then a lock would [also be held](https://docs.mongodb.com/manual/faq/concurrency/#which-administrative-commands-lock-a-collection)

Recreating the index shouldn't be an issue since it's a noop if the index already exists. So subsequent queries will use the original index.

If the docs are to be believed, I don't think a race condition exists in MongoDB 4.2 either and we can remove this comment/warning.

Reply:
*tl;dr* Yup I am satisfied that there is no race condition now having played with it.

I was concerned with the following: 
> The mongod periodically yields this lock to interleaving read and write operations.

and:

> For featureCompatibilityVersion "4.2", db.collection.createIndex() uses an optimized build process that obtains and holds an exclusive lock on the specified collection **at the start** and **end of the** index build. All subsequent operations on the collection must wait until db.collection.createIndex() releases the exclusive lock. db.collection.createIndex() allows **interleaving read and write operations during the majority of the index build**.
and: 
> MongoDB ignores the background option if specified.

In Mongo 3.6, if you create a bunch of unique data with:

`db.lock.createIndex( {"id": 1 }, { "unique": true, "background": true});` and connection B does

`db.lock.createIndex( {"id": 1 }, { "unique": true, "background": true}); db.lock.insertOne( { id: 7000000});`

Then connection A can fail with:

```
{
        "ok" : 0,
        "errmsg" : "E11000 duplicate key error collection: test.lock index: id_1 dup key: { : 7000000.0 }",
        "code" : 11000,
        "codeName" : "DuplicateKey"
}
```

But this only applies to background indexes, and it's because the createIndex() returns right away for connection B, even though in connection A it is still building. In both Mongo 3.6 and 4.4.1, if you remove background then the connection B's createIndex() blocks properly and so won't proceed to insert the data.